cmd/sdn.go

```diff
@@ -0,0 +1,686 @@
+		targetHostPort := -1
+		for k, v := range hostFwd {
+			hostPort, err := strconv.Atoi(k)
+			if err != nil {
+				log.Errorf("failed to parse host port from eve.hostfwd: %v", err)
+				continue
+			}
+			guestPort, err := strconv.Atoi(v)
+			if err != nil {
+				log.Errorf("failed to parse guest port from eve.hostfwd: %v", err)
+				continue
+			}
+			if eveIfName == "eth1" {
+				// For eth1 numbers of forwarded ports are shifted by 10.
+				hostPort += 10
+				guestPort += 10
+			}
+			if guestPort == targetPort {
+				targetHostPort = hostPort
+				break
+			}
+		}
+		if targetHostPort == -1 {
+			log.Fatalf("Target EVE interface and port (%s, %d) are not port-forwarded "+
+				"by config (see eve.hostfwd)", eveIfName, targetPort)
+		}
+		// Redirect command to localhost and the forwarded port.
+		fwdPort := strconv.Itoa(targetHostPort)
+		for i := range args {
+			args[i] = strings.ReplaceAll(args[i], fwdIPLabel, "127.0.0.1")
+			args[i] = strings.ReplaceAll(args[i], fwdPortLabel, fwdPort)
+		}
+		err := utils.RunCommandForeground(cmd, args...)
+		if err != nil {
+			log.Fatalf("Command %s failed: %v", cmd, err)
+		}
+		return
+	}
+
+	// Case 3: EVE is running inside VM on this host, with networking provided by SDN VM
+	// TODO: Port forwarding with SDN only works for TCP for now.
+
+	// Get IP address used by the target EVE interface.
+	// (look at the ARP tables inside SDN VM)
+	targetIP := getEVEIP(eveIfName)
+	if targetIP == "" {
+		log.Fatalf("no IP address found to be assigned to EVE interface %s",
+			eveIfName)
+	}
+	client := &edensdn.SdnClient{
+		SSHPort:    uint16(sdnSSHPort),
+		SSHKeyPath: sdnSSHKeyPath(),
+		MgmtPort:   uint16(sdnMgmtPort),
+	}
+	if fromEp != "" {
+		// Running command from an endpoint inside SDN VM, no tunneling is needed.
+		fwdPort := strconv.Itoa(targetPort)
+		for i := range args {
+			args[i] = strings.ReplaceAll(args[i], fwdIPLabel, targetIP)
+			args[i] = strings.ReplaceAll(args[i], fwdPortLabel, fwdPort)
+		}
+		err := client.RunCmdFromEndpoint(fromEp, cmd, args...)
+		if err != nil {
+			log.Fatalf("Command %s %s run inside endpoint %s failed: %v",
+				cmd, strings.Join(args, " "), fromEp, err)
+		}
+		return
+	}
+	// Temporarily establish port forwarding using SSH.
+	localPort, err := utils.FindUnusedPort()
+	if err != nil {
+		log.Fatalf("failed to find unused port number: %v", err)
+	}
+	closeTunnel, err := client.SSHPortForwarding(localPort, uint16(targetPort), targetIP)
+	if err != nil {
+		log.Fatalf("failed to establish SSH port forwarding: %v", err)
+	}
+	defer closeTunnel()
+	// Redirect command to localhost and the forwarded port.
+	fwdPort := strconv.Itoa(int(localPort))
+	for i := range args {
+		args[i] = strings.ReplaceAll(args[i], fwdIPLabel, "127.0.0.1")
+		args[i] = strings.ReplaceAll(args[i], fwdPortLabel, fwdPort)
+	}
+	err = utils.RunCommandForeground(cmd, args...)
+	if err != nil {
+		log.Fatalf("Command %s %s failed: %v", cmd, strings.Join(args, " "), err)
```

Comment:
log.Fatalf will call os.Exit internally and will not call deferred closeTunnel function, so it will be good to return error from the function and run Fatal from outside or to run closeTunnel here explicitly.

What is strange with tunneling (and why Eden tests in CI are failing on log_test) is the behavior on disconnection:
if I run `timeout 5 ./eden eve ssh sleep 10` without tunneling (sdn), I see (for example with `logread -f` from EVE-OS console:

```
2022-10-08T13:04:51Z,debug;Accepted publickey for root from 192.168.0.2 port 56602 ssh2: RSA SHA256:Si10Cpt9etzNitsIuDBl68C1zvJ+aGZcH3D05AlHc1A
2022-10-08T13:04:56Z,debug;Received disconnect from 192.168.0.2 port 56602:11: disconnected by user
2022-10-08T13:04:56Z,debug;Disconnected from user root 192.168.0.2 port 56602
```

but with sdn I can see only:
```
2022-10-08T13:06:18Z,debug;Accepted publickey for root from 240.0.0.2 port 39160 ssh2: RSA SHA256:TRLMCoylLTiBG+dGgwz0YIdC97GNupjbLW89bTsxRWw
```

WDYT, is it expected (in that case we should modify [log_test](https://github.com/lf-edge/eden/blob/85527ca4614620b934b2dc63cdbbda72a6a631d2/tests/lim/testdata/log_test.txt#L7)) or we do not close some sessions?

Reply:
> log.Fatalf will call os.Exit internally and will not call deferred closeTunnel function, so it will be good to return error from the function and run Fatal from outside or to run closeTunnel here explicitly.

Fixed - the function now returns error and Fatal is called from outside.

> WDYT, is it expected (in that case we should modify [log_test](https://github.com/lf-edge/eden/blob/85527ca4614620b934b2dc63cdbbda72a6a631d2/tests/lim/testdata/log_test.txt#L7)) or we do not close some sessions?

This is quite a mystery for me right now. In both cases (SDN and non-SDN) I can see full TCP termination sequence (FINs+ACKs), yet in SDN case the SSH daemon doesn't report the client being disconnected. I will continue investigating...